### PR TITLE
Exit bitcoind immediately on shutdown instead of 200ms later

### PR DESCRIPTION
**Pull request by TheBlueMatt**
This is especially useful for AbortNode(). I have no idea how to make the equivalent change in bitcoin-qt, maybe @jonasschnelli can add some signal thing for it?

### CODE REVIEW

File: src/init.cpp [hunk 1]
```diff
@@ -125,16 +125,25 @@ static const char* FEE_ESTIMATES_FILENAME="fee_estimates.dat";
...
 std::atomic<bool> fRequestShutdown(false);
 std::atomic<bool> fDumpMempoolLater(false);
+static std::mutex mutexShutdownWait;
+static std::condition_variable cvShutdownWait;
 
 void StartShutdown()
 {
     fRequestShutdown = true;
+    std::unique_lock<std::mutex> lockShutdownWait(mutexShutdownWait);
+    cvShutdownWait.notify_all();
```
Comment (JeremyRubin) [inline src/init.cpp]: Should this get mirrored to HandleSIGTERM?

File: src/init.cpp [hunk 2]
```diff
@@ -125,16 +125,25 @@ static const char* FEE_ESTIMATES_FILENAME="fee_estimates.dat";
...
 }
 bool ShutdownRequested()
 {
     return fRequestShutdown;
 }
 
+void WaitForShutdownRequested() {
+    std::unique_lock<std::mutex> lockShutdownWait(mutexShutdownWait);
+    cvShutdownWait.wait(lockShutdownWait, []() { return ShutdownRequested(); });
```
Comment (JeremyRubin) [inline src/init.cpp]: not really worth doing, but you could also wrap this in an initial `if (!ShutdownRequested()` so as not to take the lock.

**Review by JeremyRubin — COMMENTED** [review]
utack, no reason to sleep.

File: src/init.cpp (hunk 1 above)
Comment (laanwj) [inline src/init.cpp]: Are you sure this is allowed in signal handlers? From what I remember that is the reason for the circuitous polling. If not we need a special path for signals and one for the rest (like AbortNode).

**Comment by TheBlueMatt** [comment]
Made the wait still wake up every 200ms to address the sigTERM issue.

**Comment by jonasschnelli** [comment]
The QT part would require something like this (though incomplete): https://github.com/jonasschnelli/bitcoin/commit/4ed60ffcbb6e95703d6f0bf96d5f0d54e24b2fa7

**Comment by TheBlueMatt** [comment]
@jonasschnelli is there a way to do that without adding a new thread? Some kind of signal in addition to the current polling method?

**Comment by laanwj** [comment]
Yes, you could use a signal on the uiInterface that sends a Qt signal, similar to how the other signals for ClientModel work.

**Comment by jonasschnelli** [comment]
> Yes, you could use a signal on the uiInterface that sends a Qt signal, similar to how the other signals for ClientModel work.

Yes. That would be much better. I'll write a commit soon.

**Comment by laanwj** [comment]
> Yes. That would be much better. I'll write a commit soon.

Great!
I guess the uiInterface signal could be called in WaitForShutdown after the wait loop exits. This means we don't need to do any polling of fRequestShutdown ourselves in the GUI anymore. Much cleaner indeed.

File: src/init.cpp [hunk 3]
```diff
@@ -125,16 +125,27 @@ static const char* FEE_ESTIMATES_FILENAME="fee_estimates.dat";
 
 std::atomic<bool> fRequestShutdown(false);
 std::atomic<bool> fDumpMempoolLater(false);
+static std::mutex mutexShutdownWait;
+static std::condition_variable cvShutdownWait;
 
 void StartShutdown()
 {
     fRequestShutdown = true;
+    std::unique_lock<std::mutex> lockShutdownWait(mutexShutdownWait);
```
Comment (ryanofsky) [inline src/init.cpp]: Should acquire the lock before setting fRequestShutdown?

**Comment by TheBlueMatt** [comment]
Honestly I have no idea why travis if failing here. If you remove the cv notify_all() it works fine, the second you add it back travis hangs and I cannot reproduce the issue locally.

File: src/init.cpp (hunk 3 above)
Comment (TheBlueMatt) [inline src/init.cpp]: Could, doesnt matter though because fRequestShutdown is atomic.

File: src/init.cpp [hunk 4]
```diff
@@ -125,16 +125,27 @@ static const char* FEE_ESTIMATES_FILENAME="fee_estimates.dat";
 
 std::atomic<bool> fRequestShutdown(false);
 std::atomic<bool> fDumpMempoolLater(false);
+static std::mutex mutexShutdownWait;
+static std::condition_variable cvShutdownWait;
```
Comment (theuni) [inline src/init.cpp]: Use a CThreadInterrupt here instead? This is exactly what it's meant to be used for :)

**Comment by jnewbery** [comment]
I did a run of travis with the create_cache.py script commented out (so the test_runner script just launches the test cases immediately) and the tests seem to fail intermittently. Travis logs here: https://travis-ci.org/jnewbery/bitcoin/builds/217310887

Failure is:

```
raceback (most recent call last):
  File "/home/travis/build/jnewbery/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/test_framework/authproxy.py", line 125, in _request
    return self._get_response()
  File "/home/travis/build/jnewbery/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/test_framework/authproxy.py", line 167, in _get_response
    http_response = self.__conn.getresponse()
  File "/usr/lib/python3.4/http/client.py", line 1208, in getresponse
    response.begin()
  File "/usr/lib/python3.4/http/client.py", line 380, in begin
    version, status, reason = self._read_status()
  File "/usr/lib/python3.4/http/client.py", line 350, in _read_status
    raise BadStatusLine(line)
http.client.BadStatusLine: ''
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/home/travis/build/jnewbery/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/test_framework/test_framework.py", line 151, in main
    self.run_test()
  File "/home/travis/build/jnewbery/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/fundrawtransaction.py", line 449, in run_test
    self.nodes[1].encryptwallet("test")
  File "/home/travis/build/jnewbery/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/test_framework/coverage.py", line 46, in __call__
    return_val = self.auth_service_proxy_instance.__call__(*args, **kwargs)
  File "/home/travis/build/jnewbery/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/test_framework/authproxy.py", line 151, in __call__
    response = self._request('POST', self.__url.path, postdata.encode('utf-8'))
  File "/home/travis/build/jnewbery/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/test_framework/authproxy.py", line 129, in _request
    self.__conn.request(method, path, postdata, headers)
  File "/usr/lib/python3.4/http/client.py", line 1125, in request
    self._send_request(method, url, body, headers)
  File "/usr/lib/python3.4/http/client.py", line 1163, in _send_request
    self.endheaders(body)
  File "/usr/lib/python3.4/http/client.py", line 1121, in endheaders
    self._send_output(message_body)
  File "/usr/lib/python3.4/http/client.py", line 951, in _send_output
    self.send(msg)
  File "/usr/lib/python3.4/http/client.py", line 886, in send
    self.connect()
  File "/usr/lib/python3.4/http/client.py", line 863, in connect
    self.timeout, self.source_address)
  File "/usr/lib/python3.4/socket.py", line 512, in create_connection
    raise err
  File "/usr/lib/python3.4/socket.py", line 503, in create_connection
    sock.connect(sa)
ConnectionRefusedError: [Errno 111] Connection refused
```

This reproduces locally for me. Seems like there's a race between bitcoind shutting down and responding to the 'stop' RPC.

**Comment by jonasschnelli** [comment]
Maybe this Qt approach makes more sense: https://github.com/jonasschnelli/bitcoin/commit/d356512c054da926fac5a8a3f5c23721e41eaf9d

**Comment by TheBlueMatt** [comment]
@jonasschnelli cool, thanks. Sadly, in the signal handlers we can't call such a function, so we need to have both polling and a notification. I'll look into that later today.

**Comment by jonasschnelli** [comment]
@TheBlueMatt: if you want to execute the function in the main GUI thread, add another Qt signal and use Q_EMIT (emit it from the uiInterface Signal listener).

**Comment by TheBlueMatt** [comment]
Closing. I spent some time debugging further but dont think its a huge deal to sleep 200ms and it looks annoying to fix, at best. Someone else is welcome to pick this up if they feel so inclined.

**Pull request closed without merging**